### PR TITLE
Add authenticator object

### PR DESCRIPTION
**Pull request by mrodriguez90**
#### Board:
* [Ticket #NUMBER_OF_THE_TICKET](link_goes_here)
---
#### Description:

---
#### Notes:
*
---
#### Tasks:
- [x] Add each element in this format
---
#### Risk:
*
---

### CODE REVIEW

File: app/objects/impersonation/authenticator.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Impersonation
+  class Authenticator
+    def initialize(encrypted_data)
+      @encrypted_data = encrypted_data
+    end
+
+    def build_auth_headers!
+      user.tokens.filter! { |_token, attrs| attrs['impersonated_by'] != admin_user_id }
```
Comment (santib) [inline app/objects/impersonation/authenticator.rb:10]:
```suggestion
      user.tokens.reject! { |_token, attrs| attrs['impersonated_by'] == admin_user_id }
```
really minor comment, feel free to ignore :P

File: app/controllers/api/v1/impersonates_controller.rb [hunk 2]
```diff
@@ -2,13 +2,17 @@
...
-    class UsersController < API::V1::APIController
+    class ImpersonatesController < API::V1::APIController
       skip_before_action :authenticate_user!
       skip_after_action :verify_authorized
 
       def create
-        response.headers.merge!(user.build_auth_headers(token.token, token.client))
+        response.headers.merge!(
+          Impersonation::Authenticator.new(
+            params[:queryParams]
```
Comment (santib) [inline app/controllers/api/v1/impersonates_controller.rb:12]:
```suggestion
            params[:auth_enc]
```
?

File: app/objects/impersonation/authenticator.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      user.build_auth_headers(token.token, token.client) if user.save!
```
Comment (santib) [inline app/objects/impersonation/authenticator.rb:11]: `user.save!` needs to be called after `user.create_token...` right?

Comment (mrodriguez90) [reply app/objects/impersonation/authenticator.rb:11]: I've tested this locally and it seems like it's not necessary.

File: app/admin/user.rb [hunk 4]
```diff
@@ -59,6 +59,6 @@
       user_id: resource.id, admin_user_id: current_admin_user.id
     )
 
-    link_to('Impersonate User', "#{ENV.fetch('FRONTEND_URL')}?queryParams=#{encrypted_data}", method: :get)
+    link_to('Impersonate User', "#{ENV.fetch('FRONTEND_URL')}?auth_enc=#{encrypted_data}", method: :get)
```
Comment (lauperalti) [inline app/admin/user.rb:62]: The `auth_enc` name sounds weird to me, what about just `auth_encrypted` 🤔 ?

**Review by lauperalti — COMMENTED** [review]
🚀 🙌

File: app/objects/impersonation/authenticator.rb (hunk 3 above)
Comment (santib) [reply app/objects/impersonation/authenticator.rb:11]: hmm strange, is there some code saving the user after the create_token method is called then?

File: app/controllers/api/v1/impersonates_controller.rb [hunk 5]
```diff
@@ -2,13 +2,17 @@
 
 module API
   module V1
-    class UsersController < API::V1::APIController
+    class ImpersonatesController < API::V1::APIController
```
Comment (lauperalti) [inline app/controllers/api/v1/impersonates_controller.rb:5]: I think we are using `Impersonation` in other classes, does it make sense to name it Impersonation here?

**Pull request merged**